File: app/src/main/java/net/ducksmanager/whattheduck/Signup.java
package net.ducksmanager.whattheduck;


import android.content.Intent;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;

import net.ducksmanager.api.DmServer;
import net.ducksmanager.persistence.models.composite.UserToCreate;
import net.ducksmanager.util.Settings;

import java.lang.ref.WeakReference;

import androidx.appcompat.app.AppCompatActivity;
import retrofit2.Response;

public class Signup extends AppCompatActivity {

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((WhatTheDuck) getApplication()).trackActivity(this);

        setContentView(R.layout.signup);
        
        setTitle(R.string.app_name);

        EditText usernameField = findViewById(R.id.username_signup);
        EditText passwordField = findViewById(R.id.password_signup);
        EditText passwordConfirmationField = findViewById(R.id.password_confirmation);
        EditText emailField = findViewById(R.id.email_address);

        Button endSignupButton = findViewById(R.id.end_signup);
        Button cancelSignupButton = findViewById(R.id.cancel_signup);

        String defaultUsername = getIntent().getStringExtra("username");
        if (defaultUsername != null) {
            usernameField.setText(defaultUsername);
        }

        endSignupButton.setOnClickListener(view -> {
            String username = usernameField.getText().toString();
            String password = passwordField.getText().toString();

            String password2 = passwordConfirmationField.getText().toString();
            String email = emailField.getText().toString();

            DmServer.api.createUser(new UserToCreate(username, password, password2, email)).enqueue(new DmServer.Callback<Void>("signup", Signup.this, true) {
                @Override
                public void onSuccessfulResponse(Response<Void> response) {
                    DmServer.setApiDmUser(username);
                    DmServer.setApiDmPassword(Settings.toSHA1(password, new WeakReference<>(Signup.this)));
                    Login.fetchCollection(new WeakReference<>(Signup.this), CountryList.class, true);
                }
            });
        });
        
        cancelSignupButton.setOnClickListener(view -> {
            Intent i = new Intent(Signup.this, Login.class);
            Signup.this.startActivity(i);
        });
    }
}

File: app/src/main/java/net/ducksmanager/whattheduck/Signup.kt
package net.ducksmanager.whattheduck

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import androidx.appcompat.app.AppCompatActivity
import net.ducksmanager.api.DmServer
import net.ducksmanager.persistence.models.composite.UserToCreate
import net.ducksmanager.util.Settings.toSHA1
import net.ducksmanager.whattheduck.CountryList
import net.ducksmanager.whattheduck.Signup
import retrofit2.Response
import java.lang.ref.WeakReference

class Signup : AppCompatActivity() {
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (application as WhatTheDuck).trackActivity(this)
        setContentView(R.layout.signup)
        setTitle(R.string.app_name)

        val usernameField = findViewById<EditText>(R.id.username_signup)

        val defaultUsername = intent.getStringExtra("username")
        if (defaultUsername != null) {
            usernameField.setText(defaultUsername)
        }

        findViewById<Button>(R.id.end_signup)
            .setOnClickListener {
                val username = usernameField.text.toString()
                val password = findViewById<EditText>(R.id.password_signup).text.toString()
                val password2 = findViewById<EditText>(R.id.password_confirmation).text.toString()
                val email = findViewById<EditText>(R.id.email_address).text.toString()
                DmServer.api.createUser(UserToCreate(username, password, password2, email)).enqueue(object : DmServer.Callback<Void>("signup", this@Signup, true) {
                    override fun onSuccessfulResponse(response: Response<Void>) {
                        DmServer.apiDmUser = username
                        DmServer.apiDmPassword = toSHA1(password, WeakReference(this@Signup))
                        Login.Companion.fetchCollection(WeakReference<Activity>(this@Signup), CountryList::class.java, true)
                    }
                })
            }
        findViewById<Button>(R.id.cancel_signup)
            .setOnClickListener {
                this@Signup.startActivity(Intent(this@Signup, Login::class.java))
            }
    }
}